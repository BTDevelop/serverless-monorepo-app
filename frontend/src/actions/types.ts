const _REQUESTED = '_REQUESTED';
const _FULFILLED = '_FULFILLED';
const _REJECTED = '_REJECTED';

export const LOGIN_REQUESTED = 'LOGIN' + _REQUESTED;
export const LOGIN_FULFILLED = 'LOGIN' + _FULFILLED;
export const LOGIN_REJECTED = 'LOGIN' + _REJECTED;

export const LOGOUT_REQUESTED = 'LOGOUT' + _REQUESTED;
export const LOGOUT_FULFILLED = 'LOGOUT' + _FULFILLED;
export const LOGOUT_REJECTED = 'LOGOUT' + _REJECTED;

export const SYNC_AUTH_STATE_REQUESTED = 'SYNC_AUTH_STATE' + _REQUESTED;
export const SYNC_AUTH_STATE_FULFILLED = 'SYNC_AUTH_STATE' + _FULFILLED;
export const SYNC_AUTH_STATE_REJECTED = 'SYNC_AUTH_STATE' + _REJECTED;

export const API_REQUESTED = 'API' + _REQUESTED;
export const API_FULFILLED = 'API' + _FULFILLED;
export const API_REJECTED = 'API' + _REJECTED;
