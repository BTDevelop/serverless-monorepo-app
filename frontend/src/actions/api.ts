import { createAction } from 'redux-actions';
import * as types from './types';

export const apiMetaTypes = {
  dbCreate: 'dbCreate',
  email: 'email',
  file: 'file',
  private: 'private',
  public: 'public',
};

export const apiRequested = createAction(
  types.API_REQUESTED,
  undefined,
  (_, type) => ({ type }),
);
export const apiFulfilled = createAction(
  types.API_FULFILLED,
  undefined,
  (_, type) => ({ type }),
);
export const apiRejected = createAction(
  types.API_REJECTED,
  undefined,
  (_, type) => ({ type }),
);

export const emailApiRequested = (toAddress: string) =>
  apiRequested(toAddress, apiMetaTypes.email);

export const privateApiRequested = () =>
  apiRequested(undefined, apiMetaTypes.private);

export const publicApiRequested = () =>
  apiRequested(undefined, apiMetaTypes.public);

export const fileApiRequested = (fileUrl: string, key: string) =>
  apiRequested({ fileUrl, key }, apiMetaTypes.file);

export const dbCreateApiRequested = (text: string) =>
  apiRequested(text, apiMetaTypes.dbCreate);
