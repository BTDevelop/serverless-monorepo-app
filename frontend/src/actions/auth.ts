import { createAction } from 'redux-actions';
import * as types from './types';

export const loginRequested = createAction(types.LOGIN_REQUESTED);
export const loginFulfilled = createAction(types.LOGIN_FULFILLED);
export const loginRejected = createAction(types.LOGIN_REJECTED);

export const logoutRequested = createAction(types.LOGOUT_REQUESTED);
export const logoutFulfilled = createAction(types.LOGOUT_FULFILLED);
export const logoutRejected = createAction(types.LOGOUT_REJECTED);

export const syncAuthStateRequested = createAction(
  types.SYNC_AUTH_STATE_REQUESTED,
);
export const syncAuthStateFulfilled = createAction(
  types.SYNC_AUTH_STATE_FULFILLED,
);
export const syncAuthStateRejected = createAction(
  types.SYNC_AUTH_STATE_REJECTED,
);
