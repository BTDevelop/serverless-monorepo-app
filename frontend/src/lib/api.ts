const ENDPOINT = process.env.REACT_APP_API_SERVICE_ENDPOINT || '';

const PUBLIC_ENDPOINT = `${ENDPOINT}/api/public`;
const PRIVATE_ENDPOINT = `${ENDPOINT}/api/private`;

const publicApi = async () => {
  try {
    const response = await fetch(PUBLIC_ENDPOINT, {
      cache: 'no-store',
      method: 'POST',
    });

    const data = await response.json();
    // tslint:disable
    console.log('Message:', data);
  } catch (err) {
    // tslint:disable
    console.log('error', err);
  }
};

const privateApi = async () => {
  try {
    const token = localStorage.getItem('id_token');
    const response = await fetch(PRIVATE_ENDPOINT, {
      headers: {
        Authorization: `Bearer ${token}`,
      },
      method: 'POST',
    });

    const data = await response.json();
    // tslint:disable
    console.log('Token:', data);
  } catch (err) {
    // tslint:disable
    console.log('error', err);
  }
};

export { publicApi, privateApi };
