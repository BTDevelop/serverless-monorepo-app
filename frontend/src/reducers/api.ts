import { combineReducers, Reducer, Action } from 'redux';
import { handleActions } from 'redux-actions';
import { filterActions } from 'redux-ignore';
import * as actions from '../actions/api';
import * as types from '../actions/types';
import { IActionWithMeta, IApiState } from '../types/redux';

export const predicate = (action: IActionWithMeta, metaType: string) => {
  if (action.meta && action.meta.type === metaType) {
    return true;
  }
  return false;
};

const initialState: IApiState = {
  error: null,
  inProgress: false,
  result: null,
};

const apiSubReducer = handleActions<IApiState, any, any>(
  {
    [types.API_REQUESTED]: () => {
      const newState = {
        error: null,
        inProgress: true,
        result: null,
      };
      return newState;
    },
    [types.API_FULFILLED]: (_, action) => {
      const newState = {
        error: null,
        inProgress: false,
        result: action.payload,
      };
      return newState;
    },
    [types.API_REJECTED]: (_, action) => {
      const newState = {
        error: action.payload,
        inProgress: false,
        result: null,
      };
      return newState;
    },
  },
  initialState,
) as Reducer<IApiState | undefined>;

export let subReducers: {
  [metaType: string]: Reducer<IApiState | undefined>;
} = {};
Object.keys(actions.apiMetaTypes).forEach(metaType => {
  subReducers[metaType] = filterActions(apiSubReducer, (action: Action) =>
    predicate(action as IActionWithMeta, metaType),
  );
});

export default combineReducers(subReducers);
