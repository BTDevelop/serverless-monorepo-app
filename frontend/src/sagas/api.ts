import { call, put, takeEvery } from 'redux-saga/effects';
import * as actions from '../actions/api';
import * as types from '../actions/types';
import {
  createTodosItem,
  emailApi,
  privateApi,
  publicApi,
  saveFile,
} from '../lib/api';
import { IActionWithMeta } from '../types/redux';

const handlers = {
  [actions.apiMetaTypes.email]: emailApi,
  [actions.apiMetaTypes.private]: privateApi,
  [actions.apiMetaTypes.public]: publicApi,
  [actions.apiMetaTypes.file]: saveFile,
  [actions.apiMetaTypes.dbCreate]: createTodosItem,
};

export const getHandler = (type: string) => {
  return handlers[type];
};

export function* handleApi(action: IActionWithMeta) {
  const { payload, meta } = action;
  const { type } = meta;

  try {
    const handler = yield call(getHandler, type);
    const result = yield call(handler, payload);
    yield put(actions.apiFulfilled(result, type));
  } catch (e) {
    yield put(actions.apiRejected(e, type));
  }
}

export function* apiRequested() {
  yield takeEvery(types.API_REQUESTED, handleApi);
}

const sagas = [apiRequested()];

export default sagas;
