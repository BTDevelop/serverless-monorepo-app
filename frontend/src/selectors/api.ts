import { apiMetaTypes } from '../actions/api';
import { IState } from '../types/redux';

export const apiStateSelector = (state: IState, type: string) =>
  state.api[type];

export const emailApiSelector = (state: IState) =>
  apiStateSelector(state, apiMetaTypes.email);

export const privateApiSelector = (state: IState) =>
  apiStateSelector(state, apiMetaTypes.private);

export const publicApiSelector = (state: IState) =>
  apiStateSelector(state, apiMetaTypes.public);
