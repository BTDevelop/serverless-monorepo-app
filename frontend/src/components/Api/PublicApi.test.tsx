import { shallow } from 'enzyme';
import * as React from 'react';
import * as renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { publicApiRequested } from '../../actions/api';
import ConnectedPublicApi, { PublicApi } from './PublicApi';

const middlewares: any[] = [];
const mockStore = configureStore(middlewares);

jest.mock('../../selectors/api');

const props = {
  error: null,
  inProgress: false,
  publicApiRequested: jest.fn(),
  result: 'result',
};

describe('Api Component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('should render correctly', () => {
    const tree = renderer.create(<PublicApi {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should render correctly with error', () => {
    const tree = renderer
      .create(<PublicApi {...props} error={new Error('error')} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should call private api on button click', () => {
    const wrapper = shallow(<PublicApi {...props} />);

    wrapper.find('#privateApi').simulate('click');

    expect(props.publicApiRequested).toHaveBeenCalledTimes(1);
  });

  test('should connect component correctly', () => {
    const { publicApiSelector } = require('../../selectors/api');
    const mockState = {
      error: null,
      inProgress: true,
      result: 'result',
    };
    publicApiSelector.mockReturnValueOnce(mockState);
    const store = mockStore();

    // @ts-ignore wrong store type
    const wrapper = shallow(<ConnectedPublicApi store={store} />);

    expect(wrapper.props().error).toBe(mockState.error);
    expect(wrapper.props().inProgress).toBe(mockState.inProgress);
    expect(wrapper.props().result).toBe(mockState.result);
    wrapper.props().publicApiRequested();

    expect(store.getActions()).toHaveLength(1);
    expect(store.getActions()[0]).toEqual(publicApiRequested());
  });
});
