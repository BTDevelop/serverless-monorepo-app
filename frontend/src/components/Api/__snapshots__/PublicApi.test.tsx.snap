// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Api Component should render correctly 1`] = `
Array [
  <button
    className="ui button"
    disabled={undefined}
    onClick={[Function]}
    role="button"
    tabIndex={undefined}
  >
    Public Api
  </button>,
  <div
    className="ui success message"
  >
    <div
      className="content"
    >
      <div
        className="header"
      >
        Public Api Result
      </div>
      <p>
        result
      </p>
    </div>
  </div>,
]
`;

exports[`Api Component should render correctly with error 1`] = `
Array [
  <button
    className="ui button"
    disabled={undefined}
    onClick={[Function]}
    role="button"
    tabIndex={undefined}
  >
    Public Api
  </button>,
  <div
    className="ui error message"
  >
    <div
      className="content"
    >
      <div
        className="header"
      >
        Error Accessing Public Api
      </div>
      <p>
        error
      </p>
    </div>
  </div>,
]
`;
